ci-operator/step-registry/cucushift/upgrade/toversion/cucushift-upgrade-toversion-commands.sh

```diff
@@ -0,0 +1,453 @@
+    while (( try < max_retries && response != 200 )); do
+        echo "Trying #${try}"
+        response=$(https_proxy="" HTTPS_PROXY="" curl --silent --output /dev/null --write-out %"{http_code}" "https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/${algorithm}=${hash_value}/signature-1" -v)
+        (( try += 1 ))
+        sleep 60
+    done
+    set +x
+    if (( response == 200 )); then
+        echo "${TARGET} is signed" && return 0
+    else
+        echo "Seem like ${TARGET} is not signed" && return 1
+    fi
+}
+
+# Check if admin ack is required before upgrade
+function admin_ack() {
+    if (( SOURCE_MINOR_VERSION == TARGET_MINOR_VERSION )) || (( SOURCE_MINOR_VERSION < 8 )); then
+        echo "Admin ack is not required in either z-stream upgrade or 4.7 and earlier" && return
+    fi
+
+    local out; out="$(oc -n openshift-config-managed get configmap admin-gates -o json | jq -r ".data")"
+    if [[ ${out} != *"ack-4.${SOURCE_MINOR_VERSION}"* ]]; then
+        echo "Admin ack not required: ${out}" && return
+    fi
+
+    echo "Require admin ack:\n ${out}"
+    local wait_time_loop_var=0 ack_data
+
+    ack_data="$(echo "${out}" | jq -r "keys[]")"
+    for ack in ${ack_data};
+    do
+        # e.g.: ack-4.12-kube-1.26-api-removals-in-4.13
+        if [[ "${ack}" == *"ack-4.${SOURCE_MINOR_VERSION}"* ]]
+        then
+            echo "Admin ack patch data is: ${ack}"
+            oc -n openshift-config patch configmap admin-acks --patch '{"data":{"'"${ack}"'": "true"}}' --type=merge
+        fi
+    done
+    echo "Admin-acks patch gets started"
+
+    echo -e "sleep 5 min wait admin-acks patch to be valid...\n"
+    while (( wait_time_loop_var < 5 )); do
+        sleep 1m
+        echo -e "wait_time_passed=${wait_time_loop_var} min.\n"
+        if ! oc adm upgrade | grep "AdminAckRequired"; then
+            echo -e "Admin-acks patch PASSED\n"
+            return 0
+        else
+            echo -e "Admin-acks patch still in processing, waiting...\n"
+        fi
+        (( wait_time_loop_var += 1 ))
+    done
+    if (( wait_time_loop_var >= 5 )); then
+        echo >&2 "Timed out waiting for admin-acks completing, exiting" && return 1
+    fi
+}
+
+# judge if upgrade with --to or --to-latest cmd
+# set correct target_version for the upgrade
+function set_target_and_upgrade_cmd(){
+    mapfile -t available_versions < <(echo "$(oc get clusterversion version -o json|jq -r '.status.availableUpdates[]?.version')")
+    if [[ -z "${available_versions[*]}" ]]; then
+        echo "No recommended update available!"
+        exit 1
+    fi
+    # x.y.z: run upgrade with "oc adm upgrade --to x.y.z";
+    if [[ -n "${UPGRADE_TO_VERSION}" ]] && [[ "${UPGRADE_TO_VERSION}" != "latest" ]]; then
+        valid_target="false"
+        # check if x.y.z is in available_versions, if not, it's invalid target with --to.
+        for version in ${available_versions[*]}; do
```

Comment:
The checking loop looks like similar to the one on line 340, I think maybe we can extract it as a function named something like `validate_version_in_available_updates`, wdyt?

Reply:
do we really need to add one more func here for 3 lines while it is indeed in a dedicated func? and there are two difference logics here. it seems no benefit here:)

Reply:
whatever, on your call.